Add remote read filter option

grobie opened the pull request:
For special remote read endpoints which have only data for specific
queries, it is desired to limit the number of queries sent to the
configured remote read endpoint to reduce latency and performance
overhead.

This change also includes some refactoring of the remote read querier setup. The read recent and matcher filters are now small independent functions.

juliusv commented on config/config.go:
```diff
@@ -1496,6 +1496,11 @@ type RemoteReadConfig struct {
+	// Filter is an optional list of equality matcher names and values which
+	// expressions have to include in order to be sent to the remote read
+	// endpoint.
+	Filter model.LabelSet `yaml:"filter,omitempty"`
```
`filter` sounds so generic and people might think it applies to returned data and not to what remote read configs even get queried.

I haven't come up with a great alternative name yet though. Something like `query_when_matches_labels`, but much shorter :D

grobie replied:
Select this remote read endpoint when the expression has the given labels: `selector`?

grobie replied:
`condition`, `matchers`, `required_matchers`

brian-brazil replied:
They're not arbitrary matchers, so maybe `required_labels` ?

juliusv replied:
`required_labels` or `required_query_labels` (probably too long) sounds good.

grobie replied:
Changed to `required_labels`.

juliusv commented on storage/remote/read.go:
```diff
@@ -43,6 +43,48 @@ func PreferLocalFilter(cb startTimeCallback, next QuerierFunc) QuerierFunc {
+}
+
+// matcherFilterQuerier wraps a given storage.Querier, but returns NoopSeriesSet
+// for Select calls the list of matchers doesn't match the given label set.
```
the list -> where the list

juliusv commented on storage/remote/read.go:
```diff
@@ -43,6 +43,48 @@ func PreferLocalFilter(cb startTimeCallback, next QuerierFunc) QuerierFunc {
+
+// Select returns a NoopSeriesSet if the given matchers don't include equality
+// matchers for the label set of the matcherFilterQuerier. Otherwise it'll call
+// the parent querier.
```
"parent querier" doesn't feel like the right word, how about "wrapped querier"

tomwilkie commented:
(This comment is a thought-in-progress, going to go for a walk and think about it some more)

I really like the separation of the local-watermark stuff into middleware wrappers, and then the addition of this as another middleware.  Massive +1.

I'm just wondering if the the whole `QuerierFunc` mechanism is a little clunky.  Would it be 'nicer' if if was more like `http.Handler` and `http.HandlerFunc`?  ie:

```golang
type Queriable struct {
  Query(ctx context.Context, mint, maxt int64) (Querier, error)
}

type QuerierFunc func(ctx context.Context, mint, maxt int64) (Querier, error) 

func (q QuerierFunc) Query(ctx context.Context, mint, maxt int64) (Querier, error) {
  return q(ctx, mint, max)
}
```

With the `Queriable` interface embedded in `Storage` interface. `Queriable` middleware would then look like this:

```golang
func PreferLocalFilter(q Queriable) Queriable {
  return QuerierFunc(func(ctx context.Context, mint, maxt int64) (Querier, error) {
    ...
    return q.Query(ctx, min, max)
  }
}
```

And new `Storage` implementations could be constructed like this:
```golang

func PreferLocalFilterStorage(s Storage) Storage {
  return Storage{
    s,
    Querialbe: PreferLocalFilter(s.Queriable),
  }
```
...if thats even valid golang...

grobie commented:
Love the Queriable interface. Having two definitions of the query function
bugged me. Unsure about creating a storage for the filter, it isn't really
a storage or needs to be one. Gonna tinker with that once I'm home.

On Sat, Nov 11, 2017, 13:35 Tom Wilkie <notifications@github.com> wrote:

> (This comment is a thought-in-progress, going to go for a walk and think
> about it some more)
>
> I really like the separation of the local-watermark stuff into middleware
> wrappers, and then the addition of this as another middleware. Massive +1.
>
> I'm just wondering if the the whole QuerierFunc mechanism is a little
> clunky. Would it be 'nicer' if if was more like http.Handler and
> http.HandlerFunc? ie:
>
> type Queriable struct {
>   Query(ctx context.Context, mint, maxt int64) (Querier, error)
> }
> type QuerierFunc func(ctx context.Context, mint, maxt int64) (Querier, error)
>
> func (q QuerierFunc) Query(ctx context.Context, mint, maxt int64) (Querier, error) {
>   return q(ctx, mint, max)
> }
>
> With the Queriable interface embedded in Storage interface. Queriable
> middleware would then look like this:
>
> func PreferLocalFilter(q Queriable) Queriable {
>   return QuerierFunc(func(ctx context.Context, mint, maxt int64) (Querier, error) {
>     ...
>     return q.Query(ctx, min, max)
>   }
> }
>
> And new Storage implementations could be constructed like this:
>
> func PreferLocalFilterStorage(s Storage) Storage {
>   return Storage{
>     s,
>     Querialbe: PreferLocalFilter(s.Queriable),
>   }
>
> ...if thats even valid golang...
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/prometheus/prometheus/pull/3453#issuecomment-343661779>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAANaOCtMrCjzH0TBWuAlsxbEjeE3uNpks5s1ZSHgaJpZM4QZ8BS>
> .
>

grobie commented:
Thanks a lot for the suggestion @tomwilkie, this makes the QuerierFunc a lot nicer. @beorn7 mentioned to me already that the name wasn't right as the implementation didn't match the expectations people have knowing http.HandlerFunc. Somehow I didn't see how easy it was to fix this issue.

@tomwilkie @juliusv PTAL. I introduced the new interface and types to the storage interface and decoupled all the different functionalities of remote read. @fabxc you might be interested in the interface.go changes as well.

brian-brazil commented on docs/configuration/configuration.md:
```diff
@@ -1134,6 +1134,11 @@ likely in future releases.
 url: <string>
 
+# An optional list of labels which have to be matched by
+# an expression to query the remote read endpoint.
```
...to be present as equality matchers in a selector to query...

brian-brazil commented on storage/remote/read.go:
```diff
@@ -79,40 +54,124 @@ func (q *querier) Select(matchers ...*labels.Matcher) storage.SeriesSet {
+// Select returns a NoopSeriesSet if the given matchers don't match the label
+// set of the requiredLabelsQuerier. Otherwise it'll call the wrapped querier.
+func (q requiredLabelsQuerier) Select(matchers ...*labels.Matcher) storage.SeriesSet {
+	ls := q.requiredLabels
```
Is this a copy or a reference?

grobie replied:
Good catch! This function has now side effects after I changed the labels type from a slice labels.Labels to the map mode.LabelSet. Fixing

brian-brazil commented on storage/remote/read.go:
```diff
@@ -79,40 +54,124 @@ func (q *querier) Select(matchers ...*labels.Matcher) storage.SeriesSet {
+	ls := q.requiredLabels
+	for _, m := range matchers {
+		for k, v := range ls {
+			if m.Name == string(k) && m.Matches(string(v)) {
```
You need to check that it's an equality matcher

juliusv commented on storage/remote/client.go:
```diff
@@ -37,18 +37,16 @@ const maxErrMsgLen = 256
-	client     *http.Client
-	timeout    time.Duration
-	readRecent bool
+	index   int // Used to differentiate metrics.
```
Maybe unrelated but small enough fix to include: the index differentiates remote read clients, not metrics.

tomwilkie replied:
Yeah my comments not clear - its differentiates remote [read/write] clients in the metrics Prometheus exports.

grobie commented:
Please explain your reasoning to restrict the used matchers to equality matchers @brian-brazil including an example. I thought about it and removed it on purpose. We don't have any precedence for such restriction, I found it very surprising in the usage of promql and it limits the composability of remote storages.

grobie commented:
The configuration option is a set of labels which have to be machted by the expression. It's not a set of equality matchers.

brian-brazil commented:
> The configuration option is a set of labels which have to be machted by the expression. It's not a set of equality matchers.

How is a set of labels that must be the same different from a set of equality matchers?

grobie commented:
PTAL @brian-brazil @juliusv @tomwilkie. This should be ready now. Renamed `QuerierFunc` to `QueryableFunc` accordingly to the go concept, changed `required_labels` to `required_matchers` as discussed in IRC and fixed the index comment.

tomwilkie commented on storage/remote/read.go:
```diff
@@ -79,40 +54,124 @@ func (q *querier) Select(matchers ...*labels.Matcher) storage.SeriesSet {
+// before calling the wrapped storage.Queryable. The added external labels are
+// removed from the returned series sets.
+func (q externalLabelsQuerier) Select(matchers ...*labels.Matcher) storage.SeriesSet {
+	m, added := addExternalLabels(q.externalLabels, matchers)
```
Not particularly fussed, just wondering why `addExternalLabels` isn't a method on `externalLabelsQuerier`?

tomwilkie commented on storage/remote/read.go:
```diff
@@ -79,40 +54,124 @@ func (q *querier) Select(matchers ...*labels.Matcher) storage.SeriesSet {
+func (q requiredMatchersQuerier) Select(matchers ...*labels.Matcher) storage.SeriesSet {
+	ms := q.requiredMatchers
+	for _, m := range matchers {
+		for i, r := range ms {
```
Could we sort requiredMatchers and make this O(n) instead of O(n^2)?

grobie replied:
Sort both and only walk through the slices once? I guess it'll be slower in most cases, but sure, sounds good.

tomwilkie replied:
~Why would it be slower in most cases?   Just wondering if I'm missing something.~

Actually scratch that, forgot the matchers is user supplied and not ordered.  It'll be sort, so not worth the cleanup.

juliusv replied:
Yeah, it'd be very rare for more than one matcher to be supplied anyway. I think.

tomwilkie commented on storage/remote/storage.go:
```diff
@@ -86,22 +88,32 @@ func (s *Storage) ApplyConfig(conf *config.Config) error {
+			ms := make([]*labels.Matcher, 0, len(rrConf.RequiredMatchers))
+			for k, v := range rrConf.RequiredMatchers {
+				ms[i] = &labels.Matcher{Name: string(k), Value: string(v)}
+			}
```
To make this little bit of code a little cleaner, you could hide the conversion from `model.Labels` to `[]*labels.Matcher` in a function.

tomwilkie replied:
Or in the constructor for `RequiredMatchersFilter`

grobie replied:
I wouldn't want to use both model.LabelSet and labels.Matcher in the read point. It's just a log easier to use model.LabelSet in the config. I'll extract the code into a utility function.

tomwilkie commented on storage/remote/read.go:
```diff
@@ -79,40 +54,124 @@ func (q *querier) Select(matchers ...*labels.Matcher) storage.SeriesSet {
+type requiredMatchersQuerier struct {
+	storage.Querier
+
+	requiredMatchers []*labels.Matcher
```
Why is this a slice on pointers and not just a slice of structs?  What did I miss :-)?

grobie replied:
NewMatcher returns a pointer, just the result taken and put into a slice. Will change.

tomwilkie replied:
Ah I see.  Should probably use NewMatcher then.  Leave it as a pointer if its a pointer everywhere else.

grobie replied:
NewMatcher has a `(*Matcher, error)` return signature. The only error it returns is when it can't parse a regular expression. As we only need to create an equality matcher, that seemed to be unnecessary.

grobie replied:
The Querier interface is `Select(...*labels.Matcher)`. So yeah, it's pointers all over the place.

tomwilkie commented:
Looking really good!  Minor stuff from me, sorry for being slow on this one.

grobie commented:
@tomwilkie addressed your comments, except sorting and removal of pointers.

tomwilkie commented:
LGTM from me!  Thanks

brian-brazil commented:
:+1:
